// zksync-ethers/build/src/paymaster-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPaymasterParams = exports.getGeneralPaymasterInput = exports.getApprovalBasedPaymasterInput = exports.IPaymasterFlow = void 0;
const ethers_1 = require("ethers");
const utils_1 = require("./utils");
/** @deprecated This ABI is here for backward compatibility - please use utils.PAYMASTER_FLOW_ABI instead */
exports.IPaymasterFlow = new ethers_1.ethers.Interface(require("../abi/IPaymasterFlow.json").abi);
function getApprovalBasedPaymasterInput(paymasterInput) {
    return utils_1.PAYMASTER_FLOW_ABI.encodeFunctionData("approvalBased", [
        paymasterInput.token,
        paymasterInput.minimalAllowance,
        paymasterInput.innerInput,
    ]);
}
exports.getApprovalBasedPaymasterInput = getApprovalBasedPaymasterInput;
function getGeneralPaymasterInput(paymasterInput) {
    return utils_1.PAYMASTER_FLOW_ABI.encodeFunctionData("general", [paymasterInput.innerInput]);
}
exports.getGeneralPaymasterInput = getGeneralPaymasterInput;
function getPaymasterParams(paymasterAddress, paymasterInput) {
    if (paymasterInput.type == "General") {
        return {
            paymaster: paymasterAddress,
            paymasterInput: getGeneralPaymasterInput(paymasterInput),
        };
    }
    else {
        return {
            paymaster: paymasterAddress,
            paymasterInput: getApprovalBasedPaymasterInput(paymasterInput),
        };
    }
}
exports.getPaymasterParams = getPaymasterParams;
